src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Signatures/PdfSigner.cs

```diff
@@ -0,0 +1,257 @@
+﻿﻿using PdfSharp.Drawing;
+using PdfSharp.Pdf.AcroForms;
+using PdfSharp.Pdf.Annotations;
+using PdfSharp.Pdf.Internal;
+using PdfSharp.Pdf.IO;
+using System.Security.Cryptography.X509Certificates;
+
+namespace PdfSharp.Pdf.Signatures
+{
+    /// <summary>
+    /// Utility class for signing PDF-documents
+    /// </summary>
+    public class PdfSigner
+    {
+        private readonly Stream inputStream;
+
+        private readonly PdfDocument document;
+
+        private readonly ISigner signer;
+
+        private readonly PdfSignatureOptions options;
+
+        /// <summary>
+        /// Create new new instance for the specified document and with the specified options
+        /// </summary>
+        /// <param name="documentStream">Stream specifying the document to sign. Must be readable and seekable</param>
+        /// <param name="signatureOptions">The options that spefify, how the signing is performed</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="ArgumentException"></exception>
+        public PdfSigner(Stream documentStream, PdfSignatureOptions signatureOptions)
+        {
+            if (documentStream is null)
+                throw new ArgumentNullException(nameof(documentStream));
+            if (!documentStream.CanRead || !documentStream.CanSeek)
+                throw new ArgumentException("Invalid stream. Must be readable and seekable", nameof(documentStream));
+            options = signatureOptions ?? throw new ArgumentNullException(nameof(signatureOptions));
+
+            if (options.Certificate is null)
+                throw new ArgumentException("A certificate is required to sign");
+            if (options.PageIndex < 0)
+                throw new ArgumentException("Page index cannot be less than zero");
+
+            inputStream = documentStream;
+            // apply signature as an incremental update
+            document = PdfReader.Open(documentStream, PdfDocumentOpenMode.Append);
+            signer = new DefaultSigner(signatureOptions);
```

Comment:
Is there some particular reason why ISigner cannot be given in constructor? Now it always uses DefaultSIgner?

Reply:
As stated in the PR, these changes are _experimental_.
There's no real reason other than that it was the fastest way to come up with a working PoC.
Feel free to adapt to your needs.